DOC-5399: Added CPK Gradle plugin content to the docs repo (#1429)
corda/corda-docs-portal: content/en/platform/corda/5.0/developing-applications/packaging/cpk-plugin.md

---
date: '2023-06-06'
title: "CPK Gradle Plugin"
project: corda
version: 'Corda 5.0'
menu:
  corda5:
    parent: corda5-develop-packaging
    identifier: corda5-develop-packaging-cpk-plugin
    weight: 4000
section_menu: corda5
---
# CPK Gradle Plugin

Applying this plugin to a project declares that the project should create a CorDapp in CPK format. The CPK format
CorDapp is a ZIP file with a `.jar` extension, which is the output of the JAR task, along with the dependent JARs of that JAR.
The plugin will not include any of Corda's own JARs among these dependencies, nor any JARs that should be provided by
Corda, such as Kotlin or Quasar. The JAR should also contain sufficient OSGi metadata to be a valid OSGi bundle.

## Applying the Plugin

1. Depending on the number of modules of your Gradle project, choose one of the following options:

> * If your Gradle project contains multiple modules, perform the following actions:

Reviewer: Can't see a preview but it looks like there should be indentation here for the second-level bullet points to work?

Author: Corrected the indentation.

a. Include the following code at the top of the `build.gradle` file of your CorDapp Gradle project:
```
plugins {
id 'net.corda.plugins.cordapp-cpk2'
}
```
b. To configure the `cordapp-cpk2` plugin for your version of Corda, apply the `net.corda.cordapp.cordapp-configuration`
plugin to your root Gradle project:
```
plugins {
    id 'net.corda.cordapp.cordapp-configuration'
}
```

* If your Gradle project contains a single module, apply both plugins together:

```
plugins {
    id 'net.corda.cordapp.cordapp-configuration'
    id 'net.corda.plugins.cordapp-cpk2'
}
```

> 2. Declare both plugins' versions in `settings.gradle`:

Reviewer: ```suggestion
2. Declare the versions of both plugins in `settings.gradle`:
```

become transitive `cordaProvided` dependencies of any CorDapps that depend on this one.

* `cordapp`: This declares a compile-time dependency against the JAR of another CPK CorDapp. Similar to `cordaProvided`,
this dependency is also implicitly added to Gradle's `compileOnly` and `*Implementation` configurations. It is excluded
from the `runtimeClasspath` configuration, the published POM file, and the contents of the CPK file. The JARs of all
Cordapp dependencies are listed as lines in this JAR's `META-INF/CPKDependencies.json` file:

```json
{
    "formatVersion": "2.0",
    "dependencies": [
        {
            "name": "$BUNDLE_SYMBOLIC_NAME",
            "version": "$BUNDLE_VERSION",
            "verifySameSignerAsMe": true
        }
    ]
}
```

The `cordapp` dependencies are transitive, meaning that if CorDapp B declares a `cordapp` dependency on CorDapp A,
and then CorDapp C declares a cordapp dependency on CorDapp B, CorDapp C will acquire compile-time dependencies on the
JARs of both CorDapps A and B. Additionally, the `cordaProvided` dependencies of both A and B will be added to

> CorDapp C's `cordaProvided` configuration. This mechanism, is accomplished by

Reviewer: ```suggestion
CorDapp C's `cordaProvided` configuration. This mechanism is accomplished by
```

publishing each CPK with a "companion" POM that contains the additional dependency information. The `cordapp-cpk2`
plugin resolves these "companion" POMs transparently to the user, ensuring that CorDapps have the expected transitive relationships.

In order for everything to work as intended, the "companion" POM must be published into the same repository as its
associated JAR artifact. For a JAR with Maven coordinates:
```
 ${group}:${artifact}:${version}
```

> The "companion's" Maven coordinates will be:

Reviewer: ```suggestion
...then the "companion's" Maven coordinates will be:
```

```
 ${group}.${artifact}:${artifact}.corda.cpk:${version}
```

* `cordaEmbedded`: This configuration behaves similarly to `cordaProvided` in the sense that it declares a `compileOnly`
dependency that is excluded from both the CPK contents and the published POM. The difference is that the dependent JAR

> is also added to a `META-INF/lib` folder inside the CorDapp's JAR and appended to the JAR's `Bundle-Classpath` manifest

Reviewer 2: ```suggestion
is also added to a `META-INF/privatelib` folder inside the CorDapp's JAR and appended to the JAR's `Bundle-Classpath` manifest
```
